# Drink Caffeine Content Analysis Project

## Project Introdocution

In this analysis we will delve into a topic that is very near and dear to my heart - caffeine! 

I have a caffeine addiction. More often than not I opt for energy drinks ala Red Bull, Monster, Rockstar, etc. However, when the weather gets cold I start to really enjoy coffee. I often worry that the amount of caffeine I ingest will one day come back to bite me in terms of health. Despite how much caffeine I consume on a weekly basis, I know shockingly little about the amount of caffeine in the drinks I enjoy so much. 

This data set contains information related to roughly 600 different drinks with caffeine and provides information for each related to caffeine content, volume, and calories. 

## Project Goal

We will conduct some exploratory data analysis (EDA) and then attempt to find a subset of drinks which may be "less unhealthy" and could serve as some alternatives to sugar filled energy drinks and coffee.


## Import Libraries and Load the Dataset

In [51]:
# Import libraries used in the project and turn off warnings

import pandas as pd
import plotly.express as px
import plotly.io as pio
import warnings

# plotly.offline needs to be imported and the following code run in order to display visualizations offline

import plotly.offline as pyo
pyo.init_notebook_mode(connected=True)

# hiding warnings

warnings.filterwarnings('ignore')

In [52]:
# import the caffeine dataset

caffeine = pd.read_csv(r'C:\Users\benwh\Documents\Projects\Caffeine\caffeine.csv')

## Data Familiarization

In [53]:
caffeine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 610 entries, 0 to 609
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   drink          610 non-null    object 
 1   Volume (ml)    610 non-null    float64
 2   Calories       610 non-null    int64  
 3   Caffeine (mg)  610 non-null    int64  
 4   type           610 non-null    object 
dtypes: float64(1), int64(2), object(2)
memory usage: 24.0+ KB


In [54]:
caffeine.head()

drink  Volume (ml)  Calories  Caffeine (mg)    type
0                 Costa Coffee   256.993715         0            277  Coffee
1  Coffee Friend Brewed Coffee   250.191810         0            145  Coffee
2           Hell Energy Coffee   250.191810       150            100  Coffee
3           Killer Coffee (AU)   250.191810         0            430  Coffee
4                 Nescafe Gold   250.191810         0             66  Coffee

We can see after running .info() and .head() that our data set is pretty straight forward - 610 observations with five fields covering drink name, volume, calories, caffeine content and drink type.

In [55]:
# Check for duplicate drink names

duplicatedrinks = caffeine[caffeine.duplicated(['drink'])]
duplicatedrinks

Empty DataFrame
Columns: [drink, Volume (ml), Calories, Caffeine (mg), type]
Index: []

We have no duplicate drink names in our dataset.

In [56]:
# Distribution of drink types in the dataset.

caffeine['type'].value_counts().sort_values(ascending=False)

Energy Drinks    219
Coffee           173
Soft Drinks       90
Tea               66
Energy Shots      36
Water             26
Name: type, dtype: int64

Value_counts() run on the "type" field reveals there are six unique drink types.

In [57]:
# Distribution of volumes in the dataset.

caffeine['Volume (ml)'].value_counts().sort_values(ascending=False)

354.882000    159
473.176000    125
236.588000     96
250.191810     35
59.147000      19
             ... 
14.786750       1
650.617000      1
256.993715      1
751.166900      1
329.744525      1
Name: Volume (ml), Length: 70, dtype: int64

The volume of drinks varies quite a bit, but more than half of all volumes are higher than 200 mL.

In [58]:
# Distribution of calorie contnet in the dataset.
caffeine['Calories'].value_counts().sort_values(ascending=False)

0      200
5       32
10      30
160     22
140     21
      ... 
12       1
163      1
139      1
74       1
299      1
Name: Calories, Length: 98, dtype: int64

Calories per drink appear to vary quite a bit. Notably, just under half appear to be very low in calories.

In [59]:
# Distribution of caffeine content in the dataset.
caffeine['Caffeine (mg)'].value_counts().sort_values(ascending=False)

80     37
100    34
160    33
0      28
120    23
       ..
68      1
178     1
157     1
36      1
99      1
Name: Caffeine (mg), Length: 162, dtype: int64

While there are six types of drinks included in the dataset, there is a lot more variation among volume, calories, and caffeine content.

## Univariate Analysis 

### Summary Statistics for Volume (ml), Calories, and Caffeine (mg)

In [60]:
# Summary statistics for numerical data types in the dataset as a whole

caffeine.describe()

Volume (ml)    Calories  Caffeine (mg)
count   610.000000  610.000000     610.000000
mean    346.543630   75.527869     134.693443
std     143.747738   94.799919     155.362861
min       7.393375    0.000000       0.000000
25%     236.588000    0.000000      50.000000
50%     354.882000   25.000000     100.000000
75%     473.176000  140.000000     160.000000
max    1419.528000  830.000000    1555.000000

In [61]:
# Summary statistics for numerical data types individually (i.e. Volume (mL), Calories, Caffeine (mg))

caffeine.groupby(['type']).describe()

Volume (ml)                                                  \
                    count        mean         std         min         25%   
type                                                                        
Coffee              173.0  335.870855  159.716392   12.716605  236.588000   
Energy Drinks       219.0  388.971198  106.409997  236.588000  250.191810   
Energy Shots         36.0   57.742259   22.094888    7.393375   57.076855   
Soft Drinks          90.0  355.243454   41.509635  236.588000  354.882000   
Tea                  66.0  360.474080  167.002318  177.441000  236.588000   
Water                26.0  394.590111   99.702740  236.588000  343.791938   

                                             Calories              ...  \
                    50%      75%         max    count        mean  ...   
type                                                               ...   
Coffee         325.3085  473.176  1419.52800    173.0   73.497110  ...   
Energy Drinks  354.8820  473.176   751.16690    219.0   86.671233  ...   
Energy Shots    59.1470   59.147   125.98311     36.0   16.500000  ...   
Soft Drinks    354.8820  354.882   591.47000     90.0  111.111111  ...   
Tea            236.5880  473.176   946.35200     66.0   52.757576  ...   
Water          354.8820  473.176   591.47000     26.0   11.538462  ...   

                             Caffeine (mg)                                \
                  75%    max         count        mean         std   min   
type                                                                       
Coffee         120.00  830.0         173.0  200.589595  248.222165   2.0   
Energy Drinks  149.50  320.0         219.0  147.867580   76.734535   0.0   
Energy Shots    25.00  100.0          36.0  193.416667   79.535931  75.0   
Soft Drinks    160.00  320.0          90.0   33.677778   24.915961   0.0   
Tea            107.50  299.0          66.0   55.863636   39.333637   0.0   
Water            3.75  110.0          26.0   53.730769   34.060602   0.0   

                                              
                  25%    50%     75%     max  
type                                          
Coffee         100.00  145.0  200.00  1555.0  
Energy Drinks   80.00  135.0  182.50   400.0  
Energy Shots   120.00  192.5  241.25   350.0  
Soft Drinks      9.25   37.0   47.75   102.0  
Tea             29.25   45.0   68.00   165.0  
Water           35.00   60.0   73.75   120.0  

[6 rows x 24 columns]

### Distribution of Drink Types

In [62]:
caffeine_gb = caffeine.groupby(["type"]).count()
fig = px.bar(caffeine_gb, 
             x=caffeine_gb.index, 
             y="drink", 
             title='Count of Drink Types')
fig.update_layout(xaxis={'categoryorder':'total descending'})

fig.show()

### Histogram - Volume (ml)

In [63]:
fig = px.histogram(caffeine, 
             x = 'Volume (ml)', 
             title='Distribution of Volume')

fig.show()

The right tail of the histogram for Volume (ml) is pretty long. Let's zoom in and look at a histogram of volumes > 800 ml.

In [64]:
large_volume = caffeine[(caffeine['Volume (ml)'] > 600)]
large_volume

drink  Volume (ml)  Calories  Caffeine (mg)  \
29     Starbucks Bottled Iced Coffee   1419.52800       240            640   
30   Baskin Robbins Cappuccino Blast    709.76400       470            234   
31                 Dunkin' Cold Brew    709.76400         5            260   
32        Dunkin' Donuts Iced Coffee    709.76400        20            297   
33         Dunkin' Donuts Iced Latte    709.76400       100            166   
214                    Monster Hydro    751.16690       150            188   
242        Mega Monster Energy Drink    709.76400       320            240   
243         Amino Force Energy Drink    650.61700         0            200   
520           Xingtea Iced Green Tea    694.97725        50            110   
530              McDonalds Sweet Tea    946.35200       160            100   
531          Dunkin' Donuts Iced Tea    709.76400       230             63   
532                    Fuze Iced Tea    709.76400       160             24   
533                        Peace Tea    680.19050       150             23   

              type  
29          Coffee  
30          Coffee  
31          Coffee  
32          Coffee  
33          Coffee  
214  Energy Drinks  
242  Energy Drinks  
243  Energy Drinks  
520            Tea  
530            Tea  
531            Tea  
532            Tea  
533            Tea

Our large right-sided tail is made up of 13 drinks. Some of these have very large volume sizes, such as the Starbucks Bottled Iced Coffee - it's nearly 1.5 liters!

### Histogram - Calories

In [65]:
fig = px.histogram(caffeine, 
             x = 'Calories', 
             title='Distribution of Calories')

fig.show()

Similar to to Volume (ml), we have a large right-sided tail for Calories. Let's take a closer look and filter for drinks with >= 400 calories.

Also worth nothing is there are a large number of drinks with less than 20 calories.

In [66]:
high_calories = caffeine[(caffeine['Calories'] >= 400)]
high_calories

drink  Volume (ml)  Calories  Caffeine (mg)  \
13                 Dare Iced Coffee   500.087885       429            160   
30  Baskin Robbins Cappuccino Blast   709.764000       470            234   
37             Arby's Jamocha Shake   473.176000       830             12   
81   Big Train Java Chip Ice Coffee   354.882000       410             49   

      type  
13  Coffee  
30  Coffee  
37  Coffee  
81  Coffee

The long right-sided tail for calories is made up of four drinks that all look like they are a liquid version of ice cream.

### Histogram - Caffeine (mg)

In [67]:
fig = px.histogram(caffeine, 
             x = 'Caffeine (mg)', 
             title='Distribution of Caffeine', 
             template='plotly_white')

fig.show()

Again we have a long, right-sided tail. This time for Caffeine. Let's zoom in on the tail by looking at any drink with > 400 mg of caffeine.

In [68]:
high_caffeine = caffeine[(caffeine['Caffeine (mg)'] > 400)]

fig = px.histogram(high_caffeine, 
             x = 'Caffeine (mg)', 
             nbins=10, 
             title='Distribution of Drinks With Caffeine Content Higher Than 400 mg', 
             template='plotly_white')

fig.show()

### Univariate Analysis - Takeaways

- A bit over half of the dataset is made up of Enery Drink and Coffee drink types
- Most volumes fall between 200 and 475 ml
- More than half of the drinks in the dataset appear to be low calorie (< 100 calories
- Most drinks have less than 200 mg caffeine
- ~ 75% drinks have a caffeine concentration of .75 or less

## Bivariate Analysis

### Box Plot for Caffeine Across all Drink Types

In [69]:
fig = px.box(caffeine, 
            x='type', 
            y='Caffeine (mg)',
            title='Caffeine Content By Drink Type',
            labels={'type': 'Type'},
            template='plotly_white',
            hover_data=['drink', 'Volume (ml)']
            )
fig.show()

Coffee has numerous outlying drinks where caffeine content is very high. Additionally, Coffee drinks appear to be a bit higher in terms of caffeine content than Energy Drinks. However, Energy Shots are higher on average than both Coffee and Energy Drinks. 

On average, Coffee, Energy Drinks, and Energy Shots have higher caffeine content than Soft Drinks, Tea, and Water.

### Volume x Caffeine

In [70]:
fig = px.scatter(caffeine,
            x='Volume (ml)', 
            y='Caffeine (mg)',
            color='type',  
            title='Volume x Caffeine',
            height = 800,
            template='plotly_white',
            hover_data=['drink']
            )

fig.show()

The coffee outliers can be seen here as well and it looks like they are mostly among drinks of the same volume. Energy shots are clustered around 100 ml or lower but have similar caffeine content to many enery drinks and coffee. Tea, Water, and Soft Drinks appear to have lower caffeine content. 

## Box Plot for Caffeine Concentration Across all Drink Types

We're going to take another look at caffeine distribution across all drink types, but this time we'll use caffeine concentration which is equal to Caffeine (mg) / Volume (ml). This will give us a better idea of which drink types pack a caffeine punch relative to the volume of the drink. We'll also take a look at some summary statistics for the new caffeine distribution field

In [71]:
# create Caffeine Concentration field which = Caffeine (mg) / Volume (ml)

caffeine['Caffeine Concentration'] = caffeine['Caffeine (mg)'] / caffeine['Volume (ml)']

# get summary statistics for Caffeine Concentration

caffeine['Caffeine Concentration'].describe()

count    610.000000
mean       0.594908
std        1.088245
min        0.000000
25%        0.154981
50%        0.319755
75%        0.507211
max       10.482358
Name: Caffeine Concentration, dtype: float64

In [72]:
# Caffeine Concentration boxplot for all drink types

fig = px.box(caffeine,
        x='type',
        y='Caffeine Concentration',
        title='Caffeine Concentration By Drink Type',
        labels={'type': 'Type'},
        height=800,
        template='plotly_white',
        hover_data=['drink']
        )

fig.show()

As we saw just little while ago we ran caffeine['Caffeine Concentration'].describe(), the average caffeine concentration is .594 which means that on average for roughly every 1.7 mL of volume, we get 1 mg of caffeine, which is almost a 2:1 ratio. It also looks like there are some serious outliers with the max value being 10.48.

The caffeine concentration of energy shots is clearly much higher than any other drink type. Between Coffee and Energy drinks, Coffee appears to have a higher caffeine concentration. Additionally, just like with the previous box plot showing caffeine content for all drink types, there are numerous outliers for coffee.


## Healthier Options

Now that we have a better understanding of the data set and have created a caffeine concentration value for each drink, lets see if we can identify some potentially healthier options. In this data set, we only have calories as a data point which could be extrapolated to interpret whether something is healthy or not, so we will run with the assumption that lower calories = a healtheir alternative. 

Let's filter for a subset of drinks that have 100 or less calories and are not energy shots since they arent really a "drink" persay.

In [73]:
healthy_caffeine = caffeine[(caffeine['Calories'] <= 100) & (caffeine['type'] != 'Energy Shots')].sort_values(by=['Caffeine Concentration'], ascending=False).head(20)

healthy_caffeine

drink  Volume (ml)  Calories  \
85               Black Label Brewed Coffee   354.882000         0   
102                     Very Strong Coffee   354.882000         0   
92     Devils Brew Extreme Caffeine Coffee   354.882000         0   
101                       Taft Coffee (EU)   354.882000         0   
95                High Voltage Coffee (AU)   354.882000         0   
28                       Stok Coffee Shots    12.716605        10   
84                   Black Insomnia Coffee   354.882000         0   
89   Cannonball Coffee Maximum Charge (UK)   354.882000         0   
82                        Biohazard Coffee   354.882000         3   
91                       Death Wish Coffee   354.882000         0   
23                           Espresso Shot    44.360250         0   
3                       Killer Coffee (AU)   250.191810         0   
170                        Biggby Espresso    59.147000         0   
26                   Gourmesso Coffee Pods    39.924225         0   
10                         Bizzy Cold Brew    78.961245         0   
24                  Peet's Coffee Espresso    44.360250         0   
27               Nespresso Coffee Capsules    39.924225         0   
99      Stipe Miocic Extra Strength Coffee   354.882000         0   
388            Redline Xtreme Energy Drink   236.588000         0   
166                  Dripdash Kyoto Coffee   177.441000         0   

     Caffeine (mg)           type  Caffeine Concentration  
85            1555         Coffee                4.381738  
102           1350         Coffee                3.804081  
92            1325         Coffee                3.733635  
101           1182         Coffee                3.330685  
95            1150         Coffee                3.240514  
28              40         Coffee                3.145494  
84            1105         Coffee                3.113711  
89            1101         Coffee                3.102440  
82             928         Coffee                2.614954  
91             728         Coffee                2.051386  
23              77         Coffee                1.735788  
3              430         Coffee                1.718681  
170            100         Coffee                1.690703  
26              65         Coffee                1.628084  
10             125         Coffee                1.583055  
24              70         Coffee                1.577989  
27              60         Coffee                1.502847  
99             500         Coffee                1.408919  
388            316  Energy Drinks                1.335655  
166            225         Coffee                1.268027

Looking at the first 20 results, virtually everything is a type of coffee. Also, it looks like we have some drinks with very low volumes which, when combined with the drink name in some instances, appear to be some type of espresso drink. All of the values list above are well above our mean of .594 in terms of caffeine concentration.

We could end our analysis here and walk away with a list of drinks to consider drinking as alterntives, but lets do two more things before we wrap up. Let's do the following:

1. Get a list of energy drinks with the same rules applied (100 calories or less).
2. Get a combined list of tea and water drinks using the same criteria. These could be good options if we're trying to avoid enery drinks and coffee all together.

### 1. Which Energy Drinks Have the Highest Caffeine Content and Low Calories?

In [74]:
healthy_energy_drinks = caffeine[(caffeine['Calories'] <= 100) & (caffeine['type'] == 'Energy Drinks')].sort_values(by=['Caffeine Concentration'], ascending=False).head(20)

healthy_energy_drinks

drink  Volume (ml)  Calories  Caffeine (mg)  \
388       Redline Xtreme Energy Drink      236.588         0            316   
387                  Redline Princess      236.588         0            300   
348                       Hyde Xtreme      354.882         0            400   
385          PerformElite Pre-Workout      236.588         0            225   
364                     Spike Shooter      354.882        10            300   
341              Cocaine Energy Drink      354.882        90            280   
381        EBOOST Workout Crusher Mix      236.588        40            175   
314             Spike Hardcore Energy      473.176         0            350   
275                 Loud Energy Drink      473.176        10            320   
252                       Bang Energy      473.176         0            300   
294                 Raze Energy Drink      473.176         0            300   
295             Reign Total Body Fuel      473.176         0            300   
266                        GFuel Cans      473.176         0            300   
305                   Rockstar Thermo      473.176         0            300   
306                 Rockstar XDurance      473.176        10            300   
261         Celsius Heat Energy Drink      473.176        15            300   
274          Liquid Help Energy Drink      473.176         0            300   
269  Go Fast High Octane Energy Drink      473.176         0            300   
355             Meltdown Ketone Drink      354.882        18            225   
358           Octane Energy Drink Mix      354.882        25            225   

              type  Caffeine Concentration  
388  Energy Drinks                1.335655  
387  Energy Drinks                1.268027  
348  Energy Drinks                1.127135  
385  Energy Drinks                0.951020  
364  Energy Drinks                0.845351  
341  Energy Drinks                0.788995  
381  Energy Drinks                0.739682  
314  Energy Drinks                0.739682  
275  Energy Drinks                0.676281  
252  Energy Drinks                0.634014  
294  Energy Drinks                0.634014  
295  Energy Drinks                0.634014  
266  Energy Drinks                0.634014  
305  Energy Drinks                0.634014  
306  Energy Drinks                0.634014  
261  Energy Drinks                0.634014  
274  Energy Drinks                0.634014  
269  Energy Drinks                0.634014  
355  Energy Drinks                0.634014  
358  Energy Drinks                0.634014

There are some heavy hitters in this group, some packing a caffeine concentration of > 1. Still the others contains more than 200 mg of caffeine. For reference, a 12 oz Red Bull has 111 mg of caffeine.

### 2. Which Tea and Water Drinks Have the Highest Caffeine Content and Low Calories?

In [75]:
healthy_tea_water = caffeine[(caffeine['Calories'] <= 100) & ((caffeine['type'] == 'Tea') | (caffeine['type'] == 'Water'))].sort_values(by=['Caffeine Concentration'], ascending=False).head(20)

healthy_tea_water

drink  Volume (ml)  Calories  Caffeine (mg)  \
581      Zest Highly Caffeinated Tea    236.58800         0            150   
556              Fast Lane Black Tea    236.58800         0            110   
558                        HICAF Tea    236.58800         0            110   
584                 Perrier Energize    250.19181        35             99   
528          Inko's White Tea Energy    458.38925       100            165   
580              YMateina Yerba Mate    236.58800         0             80   
598             Arti Sparkling Water    354.88200         0            120   
561        Lipton Natural Energy Tea    236.58800         0             75   
545          Brew Dr Kombucha Uplift    414.02900        80            130   
554                    Cold Brew Tea    236.58800         0             70   
539   Guayaki Yerba Mate Bottled Tea    473.17600        80            140   
602      GURU Sparkling Energy Water    354.88200         0            100   
607           Yerbae Sparkling Water    354.88200         0            100   
563                       Matcha Tea    236.58800         0             64   
578                       Yellow Tea    236.58800         0             63   
544                  Xing Craft Brew    473.17600         5            120   
609        MiO Energy Water Enhancer    236.58800         0             60   
518                PG Tips Black Tea    200.50833         0             50   
562                       Lipton Tea    236.58800         0             55   
587  Caribou BOUsted Sparkling Water    340.09525         0             75   

      type  Caffeine Concentration  
581    Tea                0.634014  
556    Tea                0.464943  
558    Tea                0.464943  
584  Water                0.395696  
528    Tea                0.359956  
580    Tea                0.338141  
598  Water                0.338141  
561    Tea                0.317007  
545    Tea                0.313988  
554    Tea                0.295873  
539    Tea                0.295873  
602  Water                0.281784  
607  Water                0.281784  
563    Tea                0.270512  
578    Tea                0.266286  
544    Tea                0.253605  
609  Water                0.253605  
518    Tea                0.249366  
562    Tea                0.232472  
587  Water                0.220526

It looks like there are more than a few good options for tea and water type drinks as well. Im definitely going to do some googling to see where I can find some of these. While the associated caffeine concentrations are lower than our mean of .594, these drinks tend to have zero calories and more than a few have the same caffeine content as a 12 oz Redbull energy drink (111 mg).

## Project Conclusion

In our analysis of the Caffeine dataset, we were able to find a subset (and a subset of that subset) of healthier caffeinated drink options based on calories, caffeine amount, and caffeine concentration. In general, excluding energy shots, coffee and energy drinks tended to have the highest caffeine amounts. However, there were plenty of tea and water type drinks with caffeine amounts and caffeine concentration that rival those of standard energy drinks such as Redbull.

